File: packages/solid-fuse/src/scroll-controller.ts
import { createSignal } from "solid-js";
import { createHandle, type Handle } from "./handle";

export type ScrollController = Handle<"scrollController"> & {
  scrollOffset: () => number;
  animateTo: (offset: number, opts?: { duration?: number }) => void;
  jumpTo: (offset: number) => void;
  dispose: () => void;
};

export function createScrollController(
  opts: { initialScrollOffset?: number } = {},
): ScrollController {
  const [scrollOffset, setScrollOffset] = createSignal(
    opts.initialScrollOffset ?? 0,
  );
  const { node, call, dispose } = createHandle("scrollController", {
    ...opts,
    setScrollOffset,
  });
  return {
    node,
    scrollOffset,
    animateTo: (offset, o) => call("animateTo", { offset, ...o }),
    jumpTo: (offset) => call("jumpTo", offset),
    dispose,
  };
}

File: packages/solid-fuse/dart/lib/src/controllers/scroll_controller.dart
import 'package:flutter/material.dart';

import '../fuse_controller.dart';
import '../node.dart';

class FuseScrollController extends FuseController<ScrollController> {
  FuseScrollController(super.node);

  @override
  ScrollController create() {
    final controller = ScrollController(
      initialScrollOffset: node.double('initialScrollOffset') ?? 0,
    );
    controller.addListener(() {
      setState('scrollOffset', controller.offset);
    });
    return controller;
  }

  @override
  Future<dynamic> call(
    ScrollController object,
    String method,
    dynamic value,
  ) async {
    switch (method) {
      case 'scrollTo':
        object.jumpTo((value as num).toDouble());
      case 'animateTo':
        final map = FuseMap.from(value)!;
        object.animateTo(
          map.double('offset') ?? 0,
          duration: Duration(milliseconds: map.int('duration') ?? 300),
          curve: Curves.easeInOut,
        );
      case 'jumpTo':
        object.jumpTo((value as num).toDouble());
    }
  }

  @override
  void dispose(ScrollController object) => object.dispose();
}
